src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,202 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> current = new Node<>(null, value, null);
+        if (head == null) {
+            head = current;
+            tail = current;
+        } else {
+            tail.next = current;
+            current.prev = tail;
+            tail = current;
+            tail.next = null;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0) {
+            checkIndex(index);
+        }
+
+        Node<T> findNode = head;
+        if (findNode == null) {
+            add(value);
+            return;
+        }
+        int counter = 0;
+        int maxLength = size() + 1;
+        while (counter != index) {
+
+            if (index >= maxLength) {
+                checkIndex(index);
+                return;
+            }
+
+            findNode = findNode.next;
+            counter++;
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (findNode == null) {
+            add(value);
+            return;
+        }
+
+        if (findNode.prev == null) {
+            head = newNode;
+            head.next = findNode;
+            findNode.prev = head;
+        } else {
+            newNode.prev = findNode.prev;
+            findNode.prev.next = newNode;
+            findNode.prev = newNode;
+            newNode.next = findNode;
+        }
     }
 
-    @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> findNode = head;
+        int counter = 0;
+
+        while (findNode != null && counter != index) {
```

Comment:
**Don’t iterate your nodes from the head to tail or vice versa.**
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html